[JIT] Reuse traced fn for jit opinfos

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75917
* __->__ #76000

Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

davidberard98 committed:
[JIT] Reuse traced fn for jit opinfos

Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8873,6 +8873,9 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
                                      'test_reference_numerics_extremal_values',
                                      dtypes=(torch.complex64, torch.complex128)),
+                        # 76046
+                        DecorateInfo(unittest.skip("Skipped!"), 'TestCudaFuserOpInfo', 'test_nvfuser_correctness',
```
prefer `unittest.expectedFailure`

davidberard98 replied:
I set the consistently failing tests to `unittest.expectedFailure`, but some of the other tests are failing occasionally depending on the inputs due to not meeting the accuracy bounds (https://github.com/pytorch/pytorch/issues/76046). I left those with `unittest.skip`.

mruberry replied:
The `toleranceOverride` decorator can be applied instead of skipping the tests if you'd prefer a looser accuracy requirement

davidberard98 replied:
yes, later we're planning on deciding which ops need to be fixed & which can have a looser accuracy requirement as part of #76046.

mruberry commented:
It'd be nice if some (all?) of the skipped tests could be xfailed instead

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Reuse traced fn for jit opinfos"


Previously, jit opinfos would only run the traced function once. This is a problem for NNC and NVFuser, where the fused implementation only runs on the second invocation.

This caches the traced function and calls the cached implementation, so that subsequent calls actually perform fusion and use the fused implementation.

[ghstack-poisoned]

eellison reviewed (approved):
LGTM

davidberard98 commented:
@pytorchmergebot merge this
